rfcs/0046-module-based-model-runtime-for-aot.md

```diff
@@ -0,0 +1,348 @@
+the [Ahead-of-Time Executor](https://discuss.tvm.apache.org/t/implementing-aot-in-tvm/9206), thereby
+enabling its use from the TVM C++ Runtime.
+
+# **Motivation**
+
+The microTVM project has made significant progress towards an Ahead-of-Time Executor for compiled
+Relay models. At the time of writing, it's now possible to codegen a TIR function which executes
+Relay models that have known shapes, don't have graph-level control flow, and execute only on the
+CPU device. Right now, the C runtime is the only such runtime environment which can interact with
+this generated code. However, significant interest exists in enabling the C++ runtime to use the
+Ahead-of-Time executor.
+
+# **Guide-level explanation**
+
+Users select the AOT executor at compile time through the traditional GraphExecutor compilation flow
+(e.g. `[tvm.relay.build](http://tvm.relay.build)`) by including `--executor=aot` in the Target
+[1]. The return value of `tvm.relay.build` in this case is an `AotExecutorFactory` Module
+object. Users instantiate the AOT executor via `AotExecutorFactory` as they do with `GraphExecutor`:
+
+```bash
+ir_mod = tvm.parser.fromtext("""\
+      #[version = "0.0.5"]
+      def @main(%a : Tensor[(1, 2), uint8], %b : Tensor[(1, 2), uint8]) {
+          %0 = %a + %b;
+          %0
+      }"""
+    )
+
+with PassConfig(opt_level=3):
+  factory : AotExecutorFactory = tvm.relay.build(
+       ir_mod, "llvm -executor=aot", module_name="my_mod")
+
+aot_executor : AotExecutor = factory["my_mod"](tvm.cpu(0))
+```
+
+`AotExecutor` supports the traditional Module-Based Model Runtime Interface and can be used as a
+user normally would `GraphExecutor`:
+
+```bash
+aot_executor.set_input("a", tvm.nd.array(np.ndarray([1, 2], dtype="uint8")))
+aot_executor.set_input("b", tvm.nd.array(np.ndarray([3, 5], dtype="uint8")))
+aot_exec.run()
+output = aot_exec.get_output(0)
+assert output.asnumpy() == np.ndarray([5, 7], dtype="uint8")
+```
+
+[1] NOTE: The target string is not the final place this customization should be made. However, it's
+been the place where we've been putting runtime-related stuff. A separate RFC will split the Target
+string into Target options (which affect tuning) and runtime options.
+
+# **Reference-level explanation**
+
+Already committed to TVM is the AotExecutorCodegen. This module produces a TIR top-level function
+which invokes the Relay operators (implemented in TIR) in a correct order. An example is given
+below:
+
+```bash
+PrimFunc([input1, input2, output]) attrs={"global_symbol": "tvmgen_my_mod_run_model", "runner_function": (bool)1} {
+  // attr [(nullptr)] device_id = 0
+  // attr [(nullptr)] device_type = 1
+  tir.tvm_call_packed("tvmgen_my_mod_fused_add", input1, input2, output)
+}
+```
+
+The AotExecutor then needs to accomplish the following to meet Module-based Model Runtime Interface:
+
+1. Allocate input and output tensors as defined in the `run_model` function using the correct Device
```

Comment:
Currently AoT executor codegen does not do this. We might want to expand on this a bit.

First of all, if we are doing this -- this roughly translates into create tir.allocate nodes inside the main PrimFunc for input and output tensors.

In order to achieve this, we could do the following : 

E1) Make the main function copy the data passed into to the allocate nodes. If so, we need to perform that copy, maybe in "set_input" function ? 
E2) Expose the space allocated via tir.allocate node as a pointer to "set_input". Therefore, the user is expected either to populate the input directly there. This would avoid a copy. However, it does not align well with the current usecase of "set_input", because it is more like "get_input_space".

In this RFC, we need to make clear what is being proposed

Reply:
you mean that AOTExecutorCodegen does not emit information to determine the device, correct? I think right now it does in the sense that it hardcodes the device to (kDLCPU, 0) wherever it uses that, right?

here I'm not suggesting we modify the codegen to create tir.allocate nodes in the body of the emitted function. I'm only proposing that we emit metadata that contains shape, dtype, and device information for the expected inputs and outputs from that main func, and then defer allocation to either a runtime component (in the case of the C++ runtime) or a compile-time component (if used with the C runtime and microTVM).

The runtime changes proposed here align with E1. We could also implement `set_input_zero_copy` to align with E2. But again these are runtime changes which require that additional output from AOTExecutorCodegen to work properly.

Reply:
> I suppose the proposal here is we create a runtime wrapper around it -- which also works but with cons of not exposing these allocations to the core compiler for further optimization.

I actually think we should defer the question of how the input tensors are loaded and consumed to the runtime. You're right that AOTExecutor here is intended to be a generic runtime wrapper when there are enough resources on the system to handle this at runtime. However, consider a microTVM use case where DMA is used to copy data from e.g. a camera into an SRAM dedicated to accelerator usage. In this case, TVM should really do as much as possible to stay out of the way--the copy operation is application- or at least SoC- specific. TVM should provide pointer and sizing information so this can be handled separately. That argues for exposing some type of `get_input_tensor` function as you're mentioning here as a first-class citizen of MBMR. I think we should take that up as we move on the C Device API and USMP integrations. 

> However, Im curious to know whether it would just easier to create a copy in the main body to tir.allocate node that get translated to a device copy -- which I think has the same effect.

I think I see what you're raising here--we have to explicitly inject the input and output tensors into USMP if they are not modeled by a `tir.allocate` somewhere. We could model them without specially injecting them by wrapping the TIR top-level function in a MBMR-compatible TIR function which contains `tir.allocate` nodes, IIUC. I'm not sure how we get away from needing to do this at present, though--for `DLDevice(kDLCPU, 0)`, no such copy is needed, but we would then still need the `tir.allocate` nodes to make the input tensors visible to USMP.

Additionally, I imagine in the double-buffered input use case, the user would need to somehow communicate that space for two input buffers is needed. Not sure how to do that now outside of downsizing a memory pool.

> I would propose being explicit here that this is a runtime wrapper.
Possibly, would it be possible to add a few notes that why/ why we should not do it the codegen of main ?

Added.

Reply:
> However, consider a microTVM use case where DMA is used to copy data from e.g. a camera into an SRAM dedicated to accelerator usage. In this case, TVM should really do as much as possible to stay out of the way--the copy operation is application- or at least SoC- specific.

> Additionally, I imagine in the double-buffered input use case, the user would need to somehow communicate that space for two input buffers is needed. Not sure how to do that now outside of downsizing a memory pool.

So it is a requirement of the accelerator that it can only access certain memories and the input tensor is not originally present in a such a memory. From an AoT perspective, ideally we should provide the compiler with this information -- because we cant generally know where the input is going to be.

Therefore, when the compiler knows whether the tensor needs copying, it could insert a copy -- that gets lowered to produce an intermediary allocate node -- to be in accelerator accessible memory. Once we have that, this should ideally translate down to a loop-nest of copy that could be scheduled via double_buffer() scheduling primitives either in TE/S-TIR. This will result in creating intermediary storage in the form of allocate nodes that could be planned properly. Additionally to that, depending on who does the copy (accelerator or the host) we need to place that -- now that copy insertion is IRModule --> IRModule transformation which I think a wrapper would not be sufficient --  I would assume copies might need ending up in the device functions.

Now, I agree that it out of scope / too much to consider for this work. In the absense of that, I think letting the set_input runtime function handling the copy make sense. It is just that the short-term solution goes way against the long term solution in my mind that sparked this conversation -- however, I agree with you that there is no easy way other than inserting the copies in the host PrimFunc of host,device PrimFunc tuple if we want to do it in a more codegen'd flow.

Reply:
ah i see. yeah this makes sense. i'm wary of introducing too much complexity here particularly when user/platform intervention may be required to implement the double-buffer (e.g. if DMA is used to fill the buffer while the SoC is sleeping). it would be great to continue discussing this in a follow-on!